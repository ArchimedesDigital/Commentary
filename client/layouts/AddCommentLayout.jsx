import '../../node_modules/mdi/css/materialdesignicons.css';

import slugify from 'slugify';

AddCommentLayout = React.createClass({

    getInitialState() {
        return {
            filters: [],
            selectedLineFrom: 0,
            selectedLineTo: 0,
            contextReaderOpen: true,
            loading: false,
        };
    },

    mixins: [ReactMeteorData],

    getMeteorData() {
        const keywords = Keywords.find().fetch();
        let commenterId = null;
        if(Meteor.user() && Meteor.user().commenterId){
            commenterId = Meteor.user().commenterId;
        };
        return {
            keywords,
            commenterId,
        };
    },

    componentWillUpdate(nextProps, nextState) {
        this.handlePermissions();
    },

    updateSelectedLines(selectedLineFrom, selectedLineTo) {
        if (selectedLineFrom === null) {
            this.setState({
                selectedLineTo,
            });
        } else if (selectedLineTo === null) {
            this.setState({
                selectedLineFrom,
            });
        } else if (selectedLineTo != null && selectedLineTo != null) {
            this.setState({
                selectedLineFrom,
                selectedLineTo,
            });
        } else {
            // do nothing
        }
    },

    toggleSearchTerm(key, value) {
        let self = this,
            filters = this.state.filters;
        let keyIsInFilter = false,
            valueIsInFilter = false,
            filterValueToRemove,
            filterToRemove;

        filters.forEach(function(filter, i) {
            if (filter.key === key) {
                keyIsInFilter = true;

                filter.values.forEach(function(filterValue, j) {
                    if (filterValue._id === value._id) {
                        valueIsInFilter = true;
                        filterValueToRemove = j;
                    }
                });

                if (valueIsInFilter) {
                    filter.values.splice(filterValueToRemove, 1);
                    if (filter.values.length === 0) {
                        filterToRemove = i;
                    }
                } else {
                    if (key === 'works') {
                        filter.values = [value];
                    } else {
                        filter.values.push(value);
                    }
                }
            }
        });


        if (typeof filterToRemove !== 'undefined') {
            filters.splice(filterToRemove, 1);
        }

        if (!keyIsInFilter) {
            filters.push({
                key,
                values: [value],
            });
        }

        this.setState({
            filters,
            skip: 0,
        });
    },

    addComment(formData) {
        const that = this;

        this.setState({
            loading: true,
        });

        const filters = this.state.filters;

        let work = null;
        let subwork = null;
        filters.forEach(function(filter) {
            if (filter.key === 'work') {
                work = values[0];
            } else if (filter.key === 'subwork') {
                subwork = values[0];
            }
        });

        if (!work) {
            work = {
                title: 'Iliad',
                slug: 'iliad',
                order: 1,
            };
        }
        if (!subwork) {
            subwork = {
                title: 'Iliad',
                slug: 'iliad',
                n: 1,
            };
        }

        let lineLetter = '';
        if (this.state.selectedLineTo === 0 && this.state.selectedLineFrom > 0) {
            // checking if one line was selected
            lineLetter = this.refs.CommentLemmaSelect.state.lineLetterValue;
        }

        let referenceWorksInputObject = {};
        const referenceWorks = ReferenceWorks.find({
            slug: formData.referenceWorksValue,
        }, {
            limit: 1,
        }).fetch();

        if (referenceWorks.length) {
            referenceWorksInputObject = {
                revisionsCreated: referenceWorks[0].created,
                reference: referenceWorks[0].title,
                referenceLink: referenceWorks[0].link,
            };
        } else {
            referenceWorksInputObject = {
                revisionsCreated: new Date(),
                reference: null,
                referenceLink: null,
            };
        }

        const commenter = Commenters.find({
            _id: formData.commenterValue,
        }).fetch()[0];

        let selectedLineTo = 0;
        if (this.state.selectedLineTo === 0) {
            selectedLineTo = this.state.selectedLineFrom;
        } else {
            selectedLineTo = this.state.selectedLineTo;
        }

        this.addNewKeywordsAndIdeas(formData.keywordsValue, formData.keyideasValue, function() {
            const keywords = [];
            that.matchKeywords(formData.keywordsValue).forEach((matchedKeyword) => {
                keywords.push(matchedKeyword);
            });
            that.matchKeywords(formData.keyideasValue).forEach((matchedKeyword) => {
                keywords.push(matchedKeyword);
            });
            console.log('keywords', keywords);

            const comment = {
                work: {
                    title: work.title,
                    slug: work.slug,
                    order: work.order,
                },
                subwork: {
                    title: subwork.title,
                    n: subwork.n,
                },
                lineFrom: that.state.selectedLineFrom,
                lineTo: selectedLineTo,
                lineLetter,
                nLines: that.state.selectedLineTo - that.state.selectedLineFrom + 1,
                revisions: [{
                    title: formData.titleValue,
                    text: formData.textValue,
                    created: referenceWorksInputObject.revisionsCreated,
                    slug: slugify(formData.titleValue),
                }],
                reference: referenceWorksInputObject.reference,
                referenceLink: referenceWorksInputObject.referenceLink,
                created: new Date(),
            };

            if (commenter) {
                comment.commenters = [{
                    _id: commenter._id,
                    name: commenter.name,
                    slug: commenter.slug,
                }];
            } else {
                comment.commenters = [{}];
            }

            if (keywords) {
                comment.keywords = keywords;
            } else {
                comment.keywords = [{}];
            }

            Meteor.call('comments.insert', comment, function(error, commentId) {
                FlowRouter.go('/commentary/?_id=' + commentId);
            });

        // TODO: handle behavior after comment added (add info about success)
        });
    },

    matchKeywords(keywords) {
        const matchedKeywords = [];
        if (keywords) {
            keywords.forEach((keyword) => {
                const foundKeyword = Keywords.find({
                    title: keyword,
                }).fetch()[0];
                matchedKeywords.push(foundKeyword);
            });
        }
        return matchedKeywords;
    },

    addNewKeywordsAndIdeas(keywords, keyideas, next) {
        const that = this;
        this.addNewKeywords(keywords, 'word', function() {
            that.addNewKeywords(keyideas, 'idea', function() {
                return next();
            });
        });
    },

    addNewKeywords(keywords, type, next) {
        if (keywords) {
            const that = this;
            const insertKeywords = [];
            keywords.forEach(function(keyword) {
                foundKeyword = that.data.keywords.find(function(d) {
                    return d.title === keyword;
                });
                console.log('foundKeyword', foundKeyword, 'keyword', keyword);
                if (foundKeyword === undefined) {
                    const _keyword = {
                        title: keyword,
                        slug: slugify(keyword),
                        type,
                    };
                    insertKeywords.push(_keyword);
                }
            });
            if (insertKeywords.length > 0) {
                Meteor.call('keywords.insert', insertKeywords, function(err, data) {
                    if (err) {
                        console.log(err);
                    } else {
                        return next();
                    }
                });
            } else {
                return next();
            }
        } else {
            return next();
        }
    },

    closeContextReader() {
        this.setState({
            contextReaderOpen: false,
        });
    },

    openContextReader() {
        this.setState({
            contextReaderOpen: true,
        });
    },

    lineLetterUpdate(value) {
        this.setState({
            lineLetter: value,
        });
    },

    handlePermissions() {
        if (Roles.subscription.ready()) {
            if (!Roles.userIsInRole(Meteor.userId(), ['developer', 'admin', 'commenter'])) {
                FlowRouter.go('/');
            }
        }
    },

    handleChangeLineN(e) {
        const filters = this.state.filters;

        if (e.from > 1) {
            let lineFromInFilters = false;

            filters.forEach(function(filter, i) {
                if (filter.key === 'lineFrom') {
                    filter.values = [e.from];
                    lineFromInFilters = true;
                }
            });

            if (!lineFromInFilters) {
                filters.push({
                    key: 'lineFrom',
                    values: [e.from],
                });
            }
        } else {
            let filterToRemove;

            filters.forEach(function(filter, i) {
                if (filter.key === 'lineFrom') {
                    filterToRemove = i;
                }
            });

            if (typeof filterToRemove !== 'undefined') {
                filters.splice(filterToRemove, 1);
            }
        }

        if (e.to < 2100) {
            let lineToInFilters = false;

            filters.forEach(function(filter, i) {
                if (filter.key === 'lineTo') {
                    filter.values = [e.to];
                    lineToInFilters = true;
                }
            });

            if (!lineToInFilters) {
                filters.push({
                    key: 'lineTo',
                    values: [e.to],
                });
            }
        } else {
            let filterToRemove;

            filters.forEach(function(filter, i) {
                if (filter.key === 'lineTo') {
                    filterToRemove = i;
                }
            });

            if (typeof filterToRemove !== 'undefined') {
                filters.splice(filterToRemove, 1);
            }
        }


        this.setState({
            filters,
        });
    },

    ifReady() {
        const ready = Roles.subscription.ready();
        return ready;
    },

    render() {
        const self = this;
        const filters = this.state.filters;
        let work;
        let subwork;
        let lineFrom;
        let lineTo;

        filters.forEach(function(filter) {
            if (filter.key === 'works') {
                work = filter.values[0];
            } else if (filter.key === 'subworks') {
                subwork = filter.values[0];
            } else if (filter.key === 'lineTo') {
                lineTo = filter.values[0];
            } else if (filter.key === 'lineFrom') {
                lineFrom = filter.values[0];
            }
        });

        console.log('AddCommentLayout.filters', this.state.filters);
        console.log('AddCommentLayout.state', this.state);

        return (
            <div>
              { this.ifReady() || this.state.loading ?
                <div className="chs-layout add-comment-layout">
                  <div>
                    <Header
                      toggleSearchTerm={ this.toggleSearchTerm }
                      handleChangeLineN={ this.handleChangeLineN }
                      filters={ this.state.filters }
                      initialSearchEnabled
                    />
                    <main>
                      <div className="commentary-comments">
                        <div className="comment-group">
                          <CommentLemmaSelect
                            ref="CommentLemmaSelect"
                            selectedLineFrom={ this.state.selectedLineFrom }
                            selectedLineTo={ this.state.selectedLineTo }
                            workSlug={ work ? work.slug : 'iliad' }
                            subworkN={ subwork ? subwork.n : 1 }
                          />
                          {this.data.commenterId ?
                          <AddComment
                            selectedLineFrom={ this.state.selectedLineFrom }
                            selectedLineTo={ this.state.selectedLineTo }
                            submitForm={ this.addComment }
                            commenterId={ Meteor.user().commenterId }
                          />
                          :
                          <div className="ahcip-spinner commentary-loading">
                            <div className="double-bounce1" />
                            <div className="double-bounce2" />
                          </div> }
                          <ContextReader
                            open={ this.state.contextReaderOpen }
                            workSlug={ work ? work.slug : 'iliad' }
                            subworkN={ subwork ? subwork.n : 1 }
                            initialLineFrom={ lineFrom ? lineFrom : 1 }
                            initialLineTo={ lineTo ? lineTo : 100 }
                            selectedLineFrom={ this.state.selectedLineFrom }
                            selectedLineTo={ this.state.selectedLineTo }
                            updateSelectedLines={ this.updateSelectedLines }
                          />
                        </div>
                      </div>
                      <FilterWidget
                        filters={ filters }
                        toggleSearchTerm={ this.toggleSearchTerm }
                      />
                    </main>
                    { /* <Footer/> */ }
                  </div>
                </div>
                :
                <div className="ahcip-spinner commentary-loading full-page-spinner">
                  <div className="double-bounce1" />
                  <div className="double-bounce2" />
                </div> }
            </div>
            );
    },
});
