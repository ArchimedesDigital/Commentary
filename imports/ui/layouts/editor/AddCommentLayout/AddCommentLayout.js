import React from 'react';
import { Meteor } from 'meteor/meteor';
import { Session } from 'meteor/session';
import { Roles } from 'meteor/alanning:roles';
import { createContainer } from 'meteor/react-meteor-data';
import slugify from 'slugify';
import cookie from 'react-cookie';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

// components:
import Header from '/imports/ui/layouts/header/Header';
import FilterWidget from '/imports/ui/components/commentary/FilterWidget';
import Spinner from '/imports/ui/components/loading/Spinner';
import CommentLemmaSelect from '/imports/ui/components/editor/addComment/CommentLemmaSelect';
import AddComment from '/imports/ui/components/editor/addComment/AddComment';
import ContextPanel from '/imports/ui/layouts/commentary/ContextPanel';

// lib
import muiTheme from '/imports/lib/muiTheme';

// api
import Commenters from '/imports/api/collections/commenters';


const AddCommentLayout = React.createClass({

	propTypes: {
		ready: React.PropTypes.bool,
		isTest: React.PropTypes.bool,
	},

	getInitialState() {
		return {
			filters: [],
			selectedLineFrom: 0,
			selectedLineTo: 0,
			contextReaderOpen: true,
			loading: false,
		};
	},

	componentWillUpdate() {
		this.handlePermissions();
	},

	handlePermissions() {
		if (Roles.subscription.ready()) {
			if (!Roles.userIsInRole(Meteor.userId(), ['editor', 'admin', 'commenter'])) {
				FlowRouter.go('/');
			}
		}
	},

	// --- BEGNI LINE SELECTION --- //

	updateSelectedLines(selectedLineFrom, selectedLineTo) {
		if (selectedLineFrom === null) {
			this.setState({
				selectedLineTo,
			});
		} else if (selectedLineTo === null) {
			this.setState({
				selectedLineFrom,
			});
		} else if (selectedLineTo != null && selectedLineTo != null) {
			this.setState({
				selectedLineFrom,
				selectedLineTo,
			});
		} else {
			// do nothing
		}
	},

	toggleSearchTerm(key, value) {
		const filters = this.state.filters;
		let keyIsInFilter = false;
		let valueIsInFilter = false;
		let filterValueToRemove;
		let filterToRemove;

		filters.forEach((filter, i) => {
			if (filter.key === key) {
				keyIsInFilter = true;

				filter.values.forEach((filterValue, j) => {
					if (filterValue._id === value._id) {
						valueIsInFilter = true;
						filterValueToRemove = j;
					}
				});

				if (valueIsInFilter) {
					filter.values.splice(filterValueToRemove, 1);
					if (filter.values.length === 0) {
						filterToRemove = i;
					}
				} else if (key === 'works') {
					filters[i].values = [value];
				} else {
					filters[i].values.push(value);
				}
			}
		});


		if (typeof filterToRemove !== 'undefined') {
			filters.splice(filterToRemove, 1);
		}

		if (!keyIsInFilter) {
			filters.push({
				key,
				values: [value],
			});
		}

		this.setState({
			filters,
			skip: 0,
		});
	},

	// --- END LINE SELECTION --- //

	// --- BEGNI ADD COMMENT --- //

	addComment(formData, textValue, textRawValue) {
		this.setState({
			loading: true,
		});

		// get data for comment:
		const work = this.getWork();
		const subwork = this.getSubwork();
		const lineLetter = this.getLineLetter();
		const referenceWorks = formData.referenceWorks;
		const commenter = this.getCommenter(formData);
		const selectedLineTo = this.getSelectedLineTo();
		const token = cookie.load('loginToken');

		// need to add new keywords first, so keyword id can be added to comment:
		this.addNewKeywordsAndIdeas(formData.keywordsValue, formData.keyideasValue, () => {

			// get keywords after they were created:
			const keywords = this.getKeywords(formData);
			const revisionId = new Meteor.Collection.ObjectID();

			// create comment object to be inserted:
			const comment = {
				work: {
					title: work.title,
					slug: work.slug,
					order: work.order,
				},
				subwork: {
					title: subwork.title,
					n: subwork.n,
				},
				lineFrom: this.state.selectedLineFrom,
				lineTo: selectedLineTo,
				lineLetter,
				nLines: (selectedLineTo - this.state.selectedLineFrom) + 1,
				revisions: [{
					_id: revisionId.valueOf(),
					title: formData.titleValue,
					text: textValue,
					textRaw: textRawValue,
					created: referenceWorks ? referenceWorks.date : new Date(),
					slug: slugify(formData.titleValue),
				}],
				commenters: commenter ? [{
					_id: commenter._id,
					name: commenter.name,
					slug: commenter.slug,
				}] : [{}],
				keywords: keywords || [{}],
				referenceWorks: referenceWorks,
				tenantId: Session.get('tenantId'),
				created: new Date(),
			};

			Meteor.call('comments.insert', token, comment, (error, commentId) => {
				FlowRouter.go('/commentary', {}, {_id: commentId});
			});
		});
	},

	matchKeywords(keywords) {
		const matchedKeywords = [];
		if (keywords) {
			keywords.forEach((keyword) => {
				const foundKeyword = Keywords.findOne({
					title: keyword.label,
				});
				matchedKeywords.push(foundKeyword);
			});
		}
		return matchedKeywords;
	},

	addNewKeywordsAndIdeas(keywords, keyideas, next) {
		this.addNewKeywords(keywords, 'word', () => {
			this.addNewKeywords(keyideas, 'idea', () => next());
		});
	},

	addNewKeywords(keywords, type, next) {
		// TODO should be handled server-side
		if (keywords) {
			const newKeywordArray = [];
			keywords.forEach((keyword) => {
				const foundKeyword = Keywords.findOne({title: keyword.label});
				if (!foundKeyword) {
					const newKeyword = {
						title: keyword.label,
						slug: slugify(keyword.label),
						type,
						tenantId: Session.get('tenantId')
					};
					newKeywordArray.push(newKeyword);
				}
			});
			if (newKeywordArray.length > 0) {
				const token = cookie.load('loginToken');
				return Meteor.call('keywords.insert', token, newKeywordArray, (err) => {
					if (err) {
						console.log(err);
						return null;
					}
					return next();
				});
			}
			return next();
		}
		return next();
	},

	getWork() {
		let work = null;
		this.state.filters.forEach((filter) => {
			if (filter.key === 'work') {
				work = values[0];
			}
		});
		if (!work) {
			work = {
				title: 'Iliad',
				slug: 'iliad',
				order: 1,
			};
		}
		return work;
	},

	getSubwork() {
		let subwork = null;
		this.state.filters.forEach((filter) => {
			if (filter.key === 'subwork') {
				subwork = values[0];
			}
		});
		if (!subwork) {
			subwork = {
				title: '1',
				n: 1,
			};
		}
		return subwork;
	},

	getLineLetter() {
		let lineLetter = '';
		if (this.state.selectedLineTo === 0 && this.state.selectedLineFrom > 0) {
			lineLetter = this.commentLemmaSelect.state ? his.commentLemmaSelect.state.lineLetterValue : null;
		}
		return lineLetter;
	},

	getReferenceWorks(formData) {
		let referenceWorks = null;
		if (formData.referenceWorksValue) {
			referenceWorks = ReferenceWorks.findOne({_id: formData.referenceWorksValue.value});
		}
		return referenceWorks;
	},

	getCommenter(formData) {
		const commenter = Commenters.findOne({
			_id: formData.commenterValue.value,
		});
		return commenter;
	},

	getSelectedLineTo() {
		let selectedLineTo = 0;
		if (this.state.selectedLineTo === 0) {
			selectedLineTo = this.state.selectedLineFrom;
		} else {
			selectedLineTo = this.state.selectedLineTo;
		}
		return selectedLineTo;
	},

	getKeywords(formData) {
		const keywords = [];
		this.matchKeywords(formData.keywordsValue).forEach((matchedKeyword) => {
			keywords.push(matchedKeyword);
		});
		this.matchKeywords(formData.keyideasValue).forEach((matchedKeyword) => {
			keywords.push(matchedKeyword);
		});
		return keywords;
	},

	// --- END ADD COMMENT --- //

	closeContextReader() {
		this.setState({
			contextReaderOpen: false,
		});
	},

	openContextReader() {
		this.setState({
			contextReaderOpen: true,
		});
	},

	lineLetterUpdate(value) {
		this.setState({
			lineLetter: value,
		});
	},

	handleChangeLineN(e) {
		const filters = this.state.filters;

		if (e.from > 1) {
			let lineFromInFilters = false;

			filters.forEach((filter, i) => {
				if (filter.key === 'lineFrom') {
					filters[i].values = [e.from];
					lineFromInFilters = true;
				}
			});

			if (!lineFromInFilters) {
				filters.push({
					key: 'lineFrom',
					values: [e.from],
				});
			}
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'lineFrom') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		if (e.to < 2100) {
			let lineToInFilters = false;

			filters.forEach((filter, i) => {
				if (filter.key === 'lineTo') {
					filters[i].values = [e.to];
					lineToInFilters = true;
				}
			});

			if (!lineToInFilters) {
				filters.push({
					key: 'lineTo',
					values: [e.to],
				});
			}
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'lineTo') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}


		this.setState({
			filters,
		});
	},

	render() {
		const { isTest } = this.props;
		const { filters } = this.state;
		let work;
		let subwork;
		let lineFrom;
		let lineTo;

		filters.forEach((filter) => {
			if (filter.key === 'works') {
				work = filter.values[0];
			} else if (filter.key === 'subworks') {
				subwork = filter.values[0];
			} else if (filter.key === 'lineTo') {
				lineTo = filter.values[0];
			} else if (filter.key === 'lineFrom') {
				lineFrom = filter.values[0];
			}
		});

		return (
			<MuiThemeProvider muiTheme={getMuiTheme(muiTheme)}>
				{!this.state.loading ?
					<div className="chs-layout chs-editor-layout add-comment-layout">
						<div>
							<Header
								toggleSearchTerm={this.toggleSearchTerm}
								handleChangeLineN={this.handleChangeLineN}
								filters={this.state.filters}
								initialSearchEnabled
								addCommentPage
							/>

							{!isTest ?
								<main>
									<div className="commentary-comments">
										<div className="comment-group">
											<CommentLemmaSelect
												ref={(component) => { this.commentLemmaSelect = component; }}
												selectedLineFrom={this.state.selectedLineFrom}
												selectedLineTo={this.state.selectedLineTo}
												workSlug={work ? work.slug : 'iliad'}
												subworkN={subwork ? subwork.n : 1}
											/>

											<AddComment
												selectedLineFrom={this.state.selectedLineFrom}
												submitForm={this.addComment}
											/>

											<ContextPanel
												open={this.state.contextReaderOpen}
												workSlug={work ? work.slug : 'iliad'}
												subworkN={subwork ? subwork.n : 1}
												lineFrom={lineFrom || 1}
												selectedLineFrom={this.state.selectedLineFrom}
												selectedLineTo={this.state.selectedLineTo}
												updateSelectedLines={this.updateSelectedLines}
												editor
											/>
										</div>
									</div>

									<FilterWidget
										filters={filters}
										toggleSearchTerm={this.toggleSearchTerm}
									/>
								</main>
							: ''}
						</div>
					</div>
					:
					<Spinner fullPage />
				}
			</MuiThemeProvider>
		);
	},
});


const AddCommentLayoutContainer = (() => {
	const ready = Roles.subscription.ready();
	return {
		ready,
	};
}, AddCommentLayout);

export default AddCommentLayoutContainer;
