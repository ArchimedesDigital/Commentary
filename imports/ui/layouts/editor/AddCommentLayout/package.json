{
  "name": "AddCommentLayout",
  "version": "0.0.1",
  "private": true,
  "main": "./AddCommentLayout.js"
}
