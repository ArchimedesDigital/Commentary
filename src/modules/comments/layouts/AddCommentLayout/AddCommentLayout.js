import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { compose } from 'react-apollo';

import slugify from 'slugify';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import qs from 'qs-lite';

// components:
import Spinner from '../../../../components/loading/Spinner';

// lib
import muiTheme from '../../../../lib/muiTheme';
import Utils from '../../../../lib/utils';

// graphql
import commentsInsertMutation from '../../graphql/mutations/insert';
import AddCommentContainer from '../../containers/AddCommentContainer/AddCommentContainer';
import serializeUrn from '../../../cts/lib/serializeUrn';


const getKeywords = (formData) => {

	const keywords = [];

	formData.tagsValue.forEach((tag) => {
		const keywordCopy = {};
		for (const [key, value] of Object.entries(tag.keyword)) {
			if (key === 'isMetionedInLemma') {
				keywordCopy[key] = tag.isMentionedInLemma;
			} else if (key !== '__typename') {
				keywordCopy[key] = value;
			}
		}
		keywords.push(keywordCopy);
	});
	return keywords;
};

/*
 *	BEGIN AddCommentLayout
 */
class AddCommentLayout extends Component {

	constructor(props) {
		super(props);

		this.state = {
			filters: [],
			selectedLineFrom: 0,
			selectedLineTo: 0,
			contextReaderOpen: true,
			loading: false,
			selectedWork: ''
		};

		// methods:
		this.addComment = this.addComment.bind(this);
		this.getSubwork = this.getSubwork.bind(this);
		this.getLineLetter = this.getLineLetter.bind(this);
		this.getSelectedLineTo = this.getSelectedLineTo.bind(this);
		this.closeContextReader = this.closeContextReader.bind(this);
		this.openContextReader = this.openContextReader.bind(this);
		this.lineLetterUpdate = this.lineLetterUpdate.bind(this);
		this.updateQuery = this.updateQuery.bind(this);
		this.getChildrenContext = this.getChildrenContext.bind(this);

		let work = 'tlg001';
		this.setState({
			textNodesUrn: serializeUrn(Utils.createLemmaCitation(work, 0, 49)),
			work: work
		});
	}
	getWork() {
		let work = null;
		this.state.filters.forEach((filter) => {
			if (filter.key === 'works') {
				work = filter.values[0];
			}
		});
		if (!work) {
			work = {
				title: 'Iliad',
				slug: '001',
				order: 1,
			};
		}
		return work;
	}

	updateQuery() {
		this.setState({
			shouldUpdateQuery: false
		});
	}
	// --- BEGNI LINE SELECTION --- //


	// --- END LINE SELECTION --- //


	// --- BEGNI ADD COMMENT --- //

	addComment(formData, possibleCommenters, textValue, textRawValue) {
		this.setState({
			loading: true,
		});

		// get data for comment:
		const work = this.getWork();
		const referenceWorks = formData.referenceWorks;
		const commenters = Utils.getCommenters(formData.commenterValue, possibleCommenters);
		const selectedLineTo = this.getSelectedLineTo();
		const lemmaCitation = Utils.createLemmaCitation(work.slug, this.state.selectedLineFrom, selectedLineTo);

		// get keywords after they were created:
		const keywords = getKeywords(formData);
		const revisionId = Date.now();

		// create comment object to be inserted:
		const comment = {
			lemmaCitation: lemmaCitation,
			nLines: (selectedLineTo - this.state.selectedLineFrom) + 1,
			revisions: [{
				_id: revisionId.valueOf(),
				title: formData.titleValue,
				text: textValue,
				textRaw: textRawValue,
				created: referenceWorks ? referenceWorks.date : new Date(),
				slug: slugify(formData.titleValue),
			}],
			commenters: commenters.map(function(commenter) {
				const _commenter = JSON.parse(JSON.stringify(commenter));
				delete _commenter._id;
				delete _commenter.__typename;
				delete _commenter.avatar;
				delete _commenter.textRaw;
				return _commenter;
			}),
			keywords: keywords || [{}],
			referenceWorks: referenceWorks,
			tenantId: sessionStorage.getItem('tenantId'),
			// created: JSON.stringify(new Date()), TODO
			status: 'publish',
		};
		const that = this;
		this.props.commentInsert(comment).then((res) => {
			if (res.data.commentInsert._id) {
				const urlParams = qs.stringify({_id: res.data.commentInsert._id});
				that.props.history.push(`/commentary?${urlParams}`);
			} else {
				this.props.history.push(`/`);
			}
		});
	}

	getWorkgetWork() {
		let work = null;
		this.state.filters.forEach((filter) => {
			if (filter.key === 'works') {
				work = filter.values[0];
			}
		});
		if (!work) {
			work = {
				title: 'Iliad',
				slug: '001',
				order: 1,
			};
		}
		return work;
	}

	getSubwork() {
		let subwork = null;
		this.state.filters.forEach((filter) => {
			if (filter.key === 'subworks') {
				subwork = filter.values[0];
			}
		});
		if (!subwork) {
			subwork = {
				title: '1',
				n: 1,
			};
		}
		return subwork;
	}

	getLineLetter() {

		const { selectedLineTo, selectedLineFrom } = this.state;

		let lineLetter = '';
		if (selectedLineTo === 0 && selectedLineFrom > 0) {
			lineLetter = this.commentLemmaSelect.state ? this.commentLemmaSelect.state.lineLetterValue : null;
		}
		return lineLetter;
	}

	getSelectedLineTo() {

		const { selectedLineTo, selectedLineFrom } = this.state;

		let newSelectedLineTo = 0;
		if (selectedLineTo === 0) {
			newSelectedLineTo = selectedLineFrom;
		} else {
			newSelectedLineTo = selectedLineTo;
		}
		return newSelectedLineTo;
	}

	// --- END ADD COMMENT --- //

	closeContextReader() {
		this.setState({
			contextReaderOpen: false,
		});
	}

	openContextReader() {
		this.setState({
			contextReaderOpen: true,
		});
	}

	lineLetterUpdate(value) {
		this.setState({
			lineLetter: value,
		});
	}

	
	getChildrenContext() {
		return getMuiTheme(muiTheme);
	}
	updatetextNodesUrn(urn) {
		this.setState({textNodesUrn: urn});
	}
	render() {

		const { loading } = this.state;
		const textNodesUrn = this.state.textNodesUrn ? this.state.textNodesUrn : 'urn:cts:greekLit:tlg0012.tlg001';

		Utils.setTitle('Add Comment | The Center for Hellenic Studies Commentaries');

		return (
			<MuiThemeProvider muiTheme={this.getChildrenContext()}>
				{!loading ?
					<div className="chs-layout chs-editor-layout add-comment-layout">
						<AddCommentContainer
							textNodesUrn={textNodesUrn}
							submitForm={this.addComment}
							updatetextNodesUrn={this.updatetextNodesUrn.bind(this)} />
					</div>
					:
					<Spinner fullPage />
				}
			</MuiThemeProvider>
		);
	}
}
AddCommentLayout.propTypes = {
	commentInsert: PropTypes.func,
	history: PropTypes.object,
};

export default compose(
	commentsInsertMutation,
)(AddCommentLayout);
