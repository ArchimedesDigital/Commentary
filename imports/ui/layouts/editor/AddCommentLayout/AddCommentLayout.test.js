import React from 'react';
import renderer from 'react-test-renderer';

// component:
import AddCommentLayout from './AddCommentLayout';

describe('AddCommentLayout', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(<AddCommentLayout />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
