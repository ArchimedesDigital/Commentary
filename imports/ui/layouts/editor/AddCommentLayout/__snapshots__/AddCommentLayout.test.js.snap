// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCommentLayout renders correctly 1`] = `
<div
  className="ahcip-spinner commentary-loading full-page-spinner"
>
  <div
    className="double-bounce1"
  />
  <div
    className="double-bounce2"
  />
</div>
`;
